designs/component-upgrades.md

```diff
@@ -0,0 +1,149 @@
+
+### EKS-A controller and CRDs
+
+Both the controller and CRDs can be upgraded by just applying the new components manifest with kubectl. This should work fine for both new versions and new builds of the same version.
+
+We have to be careful and make sure that any changes to the CRDs are backwards compatible, since at this point there are already eks-a objects in the cluster. Examples of not backwards compatible changes: adding a required field, changing the type of field, making an existing field required, etc. It’s important to note that applying the manifest with the new CRD won’t fail, even if the changes are not backwards compatible. But the objects already in the cluster will become invalid, so any kind of update to those (eg. flux reconciliation) will be be rejected by the kube api server.
+
+Updates to the controller code should also take this into consideration. They always need to consider how to handle existing objects in the cluster. For example, if changing the pattern to name certain CAPI objects in the provider templates, the controller has to be able to handle both the old and the new pattern.
+
+If there is a new eks-a api version, we will need to write conversion webhooks.
+
+### Flux
+
+The process to upgrade the Flux components is very similar to the installation process, it uses the `flux` cli:
+
+1. Write the “kustomization” layer to disk to set our own Flux images
+2. Run the `bootstrap` command. This will generate the new manifest (`flux-system/gotk-components.yaml`) and commit all files to the repository
+3. Run the `reconcile` command for `flux-system` which tells Flux to pull the manifests from git and upgrade itself
+
+This process should work for both new versions and new builds of the same version.
+
+Full documentation: https://fluxcd.io/docs/installation/#bootstrap-upgrade
+
+### Cilium
+
+The process to upgrade cilium is a bit more involved: it uses helm to generate intermediate manifests and kubectl to apply them.
+
+1. Generate pre-flight check with helm and apply it with kubectl. This “pre-pulls" the new images on the nodes to avoid `ErrImagePull` errors later during the upgrade
+2. Once it’s finished, delete the pre-flight components with kubectl
+3. Generate the new Cilium manifest using the right “upgrade compatibility” options with helm (the generated manifest is different depending on the minor version jump)  and apply it with kubectl.
+
+Depending on the implementation, we might be able to avoid creating some/all of these manifests by including them in the bundle manifest.
+
+Full documentation: https://docs.cilium.io/en/v1.10/operations/upgrade/
+
+### CAPI upgrade to `v1alpha4`
+
+As mentioned in the [CAPI](#capi) section, we don’t don’t need any extra process to perform CAPI component upgrades that involve an api version change. However, there will some extra effort needed to support `v1alpha4` in the cli. Most of this will probably apply to future api version changes.
```

Comment:
```suggestion
As mentioned in the [CAPI](#capi) section, we don’t need any extra process to perform CAPI component upgrades that involve an API version change. However, there will some extra effort needed to support `v1alpha4` in the cli. Most of this will probably apply to future API version changes.
```